osquery/tables/system/windows/windows_security_products.cpp

```diff
@@ -0,0 +1,203 @@
+      provider != WSC_SECURITY_PROVIDER_ANTIVIRUS &&
+      provider != WSC_SECURITY_PROVIDER_ANTISPYWARE) {
+    VLOG(1) << "Invalid security provider code: " << provider;
+
+    return Status(E_INVALIDARG);
+  }
+
+  // Initialize can only be called once per instance, so you need to
+  // CoCreateInstance for each security product type you want to query.
+  hr = CoCreateInstance(CLSID_WSCProductList,
+                        NULL,
+                        CLSCTX_INPROC_SERVER,
+                        __uuidof(IWSCProductList),
+                        reinterpret_cast<LPVOID*>(&PtrProductList));
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to create provider instances: " << hr;
+    return Status(hr);
+  }
+
+  // Initialize the product list with the type of security product you're
+  // interested in.
+  hr = PtrProductList->Initialize(provider);
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to initialize provider: " << hr;
+    return Status(hr);
+  }
+
+  // Get the number of security products of that type.
+  hr = PtrProductList->get_Count(&ProductCount);
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to get products count: " << hr;
+    return Status(hr);
+  }
+
+  // Loop over each product, querying the specific attributes.
+  for (LONG i = 0; i < ProductCount; i++) {
+    // Get the next security product
+    hr = PtrProductList->get_Item(i, &PtrProduct);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product item: " << hr;
+      return Status(hr);
+    }
+    wsc_entry tmp;
+    tmp.provider = provider;
+
+    // Get the product name
+    hr = PtrProduct->get_ProductName(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product name: " << hr;
+      return Status(hr);
+    }
+    tmp.product_name = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the product state
+    hr = PtrProduct->get_ProductState(&ProductState);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state: " << hr;
+      return Status(hr);
+    }
+    tmp.product_state = ProductState;
+
+    // Get the remediation path for the security product
+    hr = PtrProduct->get_RemediationPath(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get remediation path: " << hr;
+      return Status(hr);
+    }
+    tmp.remediation_path = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the signature status
+    hr = PtrProduct->get_SignatureStatus(&SignatureStatus);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get signature status: " << hr;
+      return Status(hr);
+    }
+    tmp.signature_status = SignatureStatus;
+
+    // Get the state timestamp
+    hr = PtrProduct->get_ProductStateTimestamp(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state timestamp: " << hr;
+      return Status(hr);
+    }
+    tmp.product_state_timestamp = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    PtrProduct->Release();
```

Comment:
I though you do release in scope guard, don't you?

Reply:
Yes, in the case of error we have the scope guard to prevent leaking this. If we did not release it here, we would leak every product until the last one.